pages/blog/_posts/banger.mdx

```diff
@@ -3,61 +3,123 @@ heading: A Deep Dive into a Video Rendering Pipeline
+- Finally, when the render is finished, a number of tasks need to run, such as invalidating CDN cache, creating video thumbnails, or sending email to the user when the video is ready.
+
+Let's now dive into some parts of the video rendering pipeline. 
 
 ### 1. Updating the render status
 
-The first step in the render pipeline is to update the render status and set user credits on hold
+The first step in the render pipeline is to update the render status and set user credits on hold.
 
 ```js
-const updatedVideo = await step.run('Update user balance', async () => {
-  await dbConnect()
+export const renderVideo = inngest.createFunction(
+  {
+    name: 'Render video',
+    id: 'render-video',
+    cancelOn: [
+      {
+        event: 'banger/video.create',
+        match: 'data.videoId'
+      }
+    ],
+    onFailure: async ({ error, event, step }) => {
+      await dbConnect()
+
+      const isStalled = RenderStalledError.isRenderStalledError(error)
+
+       const updatedVideo = await step.run(
+        'update-video-status-to-failed',
+        () =>
+          VideoModel.findOneAndUpdate(
+            { _id: event.data.event.data.videoId },
+            {
+              $set: {
+                status: isStalled ? 'stalled' : 'error',
+                ...(isStalled && { stalledAt: new Date() }),
+                renderProgress: null
+              }
+            },
+            { new: true }
+          )
+            .lean()
+      )
+
+      invariant(updatedVideo, 'no video found')
+
+      // refund user units if error is not recoverable
+      // if it's stalled, we're going to recover it later
+      if (!isStalled) {
+        await step.run('refund-user-units', async () => {
+          await UserModel.updateOne(
+            {
+              _id: event.data.event.data.userId
+            },
+            { $inc: { unitsRemaining: updatedVideo.videoDuration } }
+          )
+        })
+      }
 
-  const render = await VideoModel.findOneAndUpdate(
-    { _id: videoId },
-    { $set: { renderProgress: 0, renderTime: 0, status: 'pending' } },
-    { new: true }
-  )
-    .populate('user')
-    .lean()
-
-  invariant(video, 'no render found')
-
-  // Simplified
-  await UserModel.updateOne(
-    {
-      _id: video.user._id
-    },
-    {
-      $inc: { unitsRemaining: -video.videoDuration }
-    }
-  )
+      if (process.env.NODE_ENV === 'production') {
+        const errorJson = _.truncate(JSON.stringify(event), {
+          length: 3000
+        })
+        await sendTelegramLog(
+          _.truncate(
+            `🚨 Error while rendering video: ${error.message}\n
+          Event: ${errorJson}\n`,
+            { length: 3000 }
+          )
+        )
+      }
 
-  return video
-})
+      Sentry.captureException(error)
+    }
+  },
+  { event: 'banger/video.create' },
+  async ({ event, step, attempt, logger }) => {
+    // ...
```

Comment:
The old code which updated the user status (the name of this section) is now missing, is this correct? Do we mean to bring back that first step?

The `onFailure` handler is cool, but perhaps we should just move that to the end of the post in it's own section and link to our docs on `onFailure`.